pkg/syncer/cnsoperator/controller/cnsunregistervolume/cnsunregistervolume_controller.go

```diff
@@ -177,7 +181,198 @@ func (r *ReconcileCnsUnregisterVolume) Reconcile(ctx context.Context,
 	log.Infof("Reconciling CnsUnregisterVolume instance %q from namespace %q. timeout %q seconds",
 		instance.Name, request.Namespace, timeout)
 
-	//TODO - Add implementation logic for the reconciler
+	// 1. Perform all the necessary validations.
+	// 2. Fetch the PV corresponding to the volume and set on it the ReclaimPolicy to Retain.
+	// 3. Delete PVC, wait for it to get deleted.
+	// 4. Delete PV.
+	// 5. Invoke CNS DeleteVolume API with deleteDisk set to false.
+	// 6. Set the CnsUnregisterVolumeStatus.Unregistered to true.
 
+	// TODO - Add validations whether the volume is not in use in a TKC or by a VM service VM.
+
+	queryFilter := cnstypes.CnsQueryFilter{
+		VolumeIds: []cnstypes.CnsVolumeId{{Id: instance.Spec.VolumeID}},
+	}
+	querySelection := cnstypes.CnsQuerySelection{
+		Names: []string{
+			metadata,
+		},
+	}
+
+	queryResult, err := r.volumeManager.QueryVolumeAsync(ctx, queryFilter, &querySelection)
+	if err != nil {
+		msg := fmt.Sprintf("Unable to query volume %q . Error: %+v", instance.Spec.VolumeID, err)
+		log.Error(msg)
+		setInstanceError(ctx, r, instance, msg)
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+	if len(queryResult.Volumes) == 0 {
+		log.Infof("Volume: %q not found while querying CNS", instance.Spec.VolumeID)
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+
+	cnsVol := queryResult.Volumes[0]
+
+	var pvName, pvcName, namespace string
+	for _, entity := range cnsVol.Metadata.EntityMetadata {
+		entityType := entity.(*cnstypes.CnsKubernetesEntityMetadata).EntityType
+		if entityType == string(cnstypes.CnsKubernetesEntityTypePV) {
+			pvName = entity.(*cnstypes.CnsKubernetesEntityMetadata).EntityName
+		}
+
+		if entityType == string(cnstypes.CnsKubernetesEntityTypePVC) {
+			pvcName = entity.(*cnstypes.CnsKubernetesEntityMetadata).EntityName
+			namespace = entity.(*cnstypes.CnsKubernetesEntityMetadata).Namespace
+		}
+	}
+
+	if pvName == "" {
+		err := fmt.Errorf("CNS metadata for volume %s has missing pvName", instance.Spec.VolumeID)
+		log.Error(err)
+		return reconcile.Result{}, err
+	}
```

Comment:
As mentioned, there is clearly a CNS entry, but it doesn't point to a PV/PVC. This doesn't necessarily mean the PV/PVC doesn't exist. If we want this to be robust, we need to ensure all of them are deleted independently.

Reply:
As discussed, if CNS metadata doesn't have PV/PVC details then we can't really process PV/PVC and they will be taken care of in next CSI full sync.